### PR TITLE
Add Mage-VL image and video support

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

Adds first-class support for [`microsoft/Mage-VL`](https://huggingface.co/microsoft/Mage-VL) (`model_type: mage_vl`) as a standardized three-model package:

- `decoder`: Qwen3 text decoder with ordinary 1D RoPE
- `vision_encoder`: custom 24-layer Mage-ViT for images and streaming video
- `embedding`: batch-safe text/visual feature mixer

The vision graph matches Mage-VL's fused biased QKV layout, exact 4:6:6 temporal/height/width RoPE, original sampled-frame temporal positions, 2x2 patch merger, and independent four-frame attention windows. CUDA/DML builds use packed attention with dynamically derived window boundaries; portable builds retain a standard-ONNX block-diagonal fallback.

This also adds remote config extraction, registry/public exports, exact checkpoint weight routing, ordered mixed image/video processing, ORT GenAI compatibility guardrails, real media fixtures, and L1-L5 coverage.

## Validation

- Unsupported status verified before implementation; no existing `mage_vl` registry/config/model support was present.
- Upstream revision studied and pinned: `d88b153285f1633a61b2f693c59c8576693af185`.
- Real processor path exercised with a nonzero Apache-2.0 image and soccer video: 4,284 packed patches, five sampled frames, original temporal indices `[0, 180, 360, 539, 719]`, and 1,138 prompt tokens.
- `python -m pytest tests/build_graph_test.py tests/weight_alignment_test.py tests/mage_vl_parity_test.py -k "mage_vl" -q`: **10 passed**.
  - Includes exact checkpoint-key alignment, FP32 parity (`atol/rtol=1e-4`), FP16 parity (`1e-2`), four-frame-boundary video coverage, empty-media decode, multi-row media ordering, and packed-CUDA graph coverage.
- `python -m pytest tests/yaml_schema_test.py src/mobius/_testing/golden_test.py src/mobius/integrations/ort_genai/auto_export_test.py -q`: **345 passed**.
- `python -m pytest tests/build_graph_test.py tests/cli_test.py src/ -q -k "not phi4mm and not apply_weights_unknown" --tb=short -n auto`: **3,639 passed, 57 skipped**.
- `python -m pytest tests/quantization_integration_test.py -v`: **6 passed**.
- `lintrunner -a`: **clean**.
- Automated code review was run repeatedly; findings for batch indexing, strict image-only processors, and quadratic video attention were fixed. Final review reported no significant issues.

### Real checkpoint / GPU

Hardware: NVIDIA RTX A1000, 8,188 MiB; ORT 1.28.0 CUDA EP.

- BF16 L4 mixed real image/video, CUDA-specialized build and `CUDAExecutionProvider`: **1 passed in 324.12s**.
- BF16 L5 deterministic generation, 24 new tokens, same real image/video and CUDA path: **1 passed in 236.19s**.
- The committed HF golden's 24-token continuation begins: `The image depicts a group of four men standing around a table in a stadium...`.
- No peak-memory number is claimed.

### CLI exports

Direct three-model ONNX export remains supported for CPU and CUDA builds. `--runtime ort-genai` is explicitly rejected before any model artifacts are saved because ORT GenAI cannot supply Mage-VL's required `patch_positions` vision input or its 1D decoder `position_ids`.

```text
Error: ORT GenAI does not support Mage-VL's required patch_positions vision input or its 1D decoder position_ids contract. Export without --runtime ort-genai to save the runnable direct three-model ONNX package.
```

### Olive

Olive 0.10.2 NF4 quantization completed on the real FP16 decoder in **74.46s**, reducing external weights from **8,112,046,080 bytes** to **2,204,314,624 bytes**. The quantized decoder loaded in ORT CPU, produced finite `(1, 3, 151936)` logits, and generated:

```text
The capital of France is Paris. The capital of Germany is Berlin
```

## Known runtime limits

- **ORT GenAI / Foundry Local:** unsupported until the runtime supports both Mage-VL's required `patch_positions` input (including original sampled-frame indices) and Mage-VL's 1D decoder positions. Mobius rejects `--runtime ort-genai` instead of emitting a non-runnable package.
- **Direct ONNX:** supported as the standardized `decoder` + `vision_encoder` + `embedding` package and covered by real CUDA L4/L5 validation.
- **Full real-weight FP32/FP16 generation matrix:** FP32 and FP16 are covered by independent synthetic numerical parity; FP16 is additionally covered by full real-weight direct ONNX exports; native BF16 is covered by real CUDA L4/L5. A full FP32 checkpoint run would approximately double the 9.5 GB BF16 checkpoint footprint and was not repeated on this 8 GB GPU.

### CODE REVIEW

**Comment by justinchuby** [comment]
## CI triage update

The first CI run on `05b908f` had two distinct results:

- **Mage-VL:** L1 passed. The general Ubuntu matrix found a missing L2 registry metadata entry (`test_model_id`), which is Mage-specific; commit `3eda7ff` adds `microsoft/Mage-VL` and the focused coverage plus real remote-config graph validation now pass locally (`4 passed` and `3 passed`, respectively).
- **Bamba:** L3 and all three Ubuntu unit-test matrices reported only the same unrelated pre-existing numerical-threshold failures, with unchanged argmax/top-10 behavior:
  - `bamba_0`: max abs `0.001375`, mean abs `0.000201`, cosine `0.999997`, argmax match `True`
  - `bamba_1`: max abs `0.001662`, mean abs `0.000172`, cosine `0.999996`, argmax match `True`

No Bamba source, configuration, tolerance, or test code is changed by this PR, and I will not modify unrelated Bamba behavior to make this PR green.

Baseline reference: the PR base is `78437cd`; its main-branch CI run [31230202537](https://github.com/onnxruntime/mobius/actions/runs/31230202537) already fails overall (L4/L5). For precision, that baseline run's L3 job itself passed, so the Bamba result is treated as an unrelated numerical flake rather than claimed as an identical deterministic failure in that specific baseline L3 execution.

The follow-up CI run for `3eda7ff` is in progress. Mage-specific L1/L4/L5 results will be recorded when it completes.

**Comment by justinchuby** [comment]
## Follow-up CI result (`3eda7ff`)

The Mage-specific L2 metadata fix is effective:

- Focused registry coverage: **4 passed**.
- L2 remote config + full-size graph validation for `microsoft/Mage-VL`: **3 passed**.
- Follow-up CI L1 Smoke Test: **passed**.
- All six Linux/Windows unit matrices now report **5,039 passed** and only the same two Bamba failures; the prior Mage registry-coverage failures are gone.
- Follow-up L3 again reports **100 passed**, with only `bamba_0` and `bamba_1` failing at the exact metrics documented above. No Bamba changes were made.

The self-hosted A10 L4/L5 jobs are still queued with no runner assigned, so I am not claiming those CI jobs ran. Mage-specific L4/L5 are nevertheless independently verified by the real nonzero image+five-frame video CUDA runs already detailed in the PR body: BF16 L4 **passed in 324.12s** and deterministic 24-token BF16 L5 **passed in 236.19s** on `CUDAExecutionProvider`.

File: src/mobius/integrations/ort_genai/auto_export.py [hunk 1]
```diff
@@ -846,13 +863,25 @@ def _write_genai_config(
...
+            model_type = getattr(config, "model_type", "")
             vision_input_mapping = _introspect_inputs(pkg, "vision_encoder")
             embedding_input_mapping = _introspect_inputs(pkg, "embedding")
+            if model_type == "mage_vl" and vision_input_mapping is not None:
+                # ORT GenAI 0.15 rejects unknown vision input mapping keys while
+                # Mage-VL requires per-patch sampled-frame positions. Keep the
+                # config schema-loadable for inspection; callers must run the
+                # standardized three-model package directly until the runtime
+                # supports both this input and Mage-VL's 1D decoder positions.
+                vision_input_mapping.pop("patch_positions", None)
```
Comment (justinchuby) [inline src/mobius/integrations/ort_genai/auto_export.py]: Removing `patch_positions` only from the runtime mapping leaves the vision ONNX graph with a required input that ORT GenAI never supplies, so `--runtime ort-genai` emits a schema-loadable but non-runnable package. A warning does not preserve the exporter contract. Please reject Mage-VL ORT GenAI export explicitly until the runtime can generate/map this input (and 1D decoder positions), or implement the missing processor/runtime support.

**Review by justinchuby — COMMENTED** [review]
Correctness review: one runtime-contract finding.

File: src/mobius/integrations/ort_genai/auto_export.py (hunk 1 above)
Comment (justinchuby) [reply src/mobius/integrations/ort_genai/auto_export.py]: Fixed in d5870c6. Mage-VL is removed from ORT GenAI model routing and all runtime export entry points now reject it before saving ONNX/config artifacts, with CLI/API regression tests. Direct three-model ONNX export remains supported.

File: tests/e2e_golden_test.py [hunk 2]
```diff
@@ -834,20 +834,92 @@ def _run_vl_vision_to_image_features(
...
+            if video_processor is not None:
+                video_processor.max_pixels = case.media_max_pixels
+        processed_pt = processor(**kwargs)
+    finally:
+        if image_processor is not None and saved_image_max is not None:
+            image_processor.max_pixels = saved_image_max
+        if image_size is not None and saved_image_longest is not None:
+            image_size.longest_edge = saved_image_longest
+        if video_processor is not None and saved_video_max is not None:
+            video_processor.max_pixels = saved_video_max
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/e2e_golden_test.py]: Fixed in d5870c6. Processor pixel-limit overrides now use a shared context manager that records exact state, restoring `None` values and deleting attributes that were originally absent. Added focused regression coverage.

File: scripts/generate_golden.py [hunk 3]
```diff
@@ -451,11 +454,37 @@ def _generate_vision_language(case: TestCase, json_path: Path, device: str) -> N
...
+            processor_kwargs["videos"] = videos
+            processor_kwargs["num_frames"] = case.video_num_frames
+        processed = processor(**processor_kwargs)
+    finally:
+        if image_processor is not None and saved_image_max is not None:
+            image_processor.max_pixels = saved_image_max
+        if image_size is not None and saved_image_longest is not None:
+            image_size.longest_edge = saved_image_longest
+        if video_processor is not None and saved_video_max is not None:
+            video_processor.max_pixels = saved_video_max
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply scripts/generate_golden.py]: Fixed in d5870c6. Golden generation now uses `_temporary_processor_max_pixels`, which restores each image/video processor attribute exactly, including `None` and missing attributes, in a `finally` block.

File: tests/cli_test.py [hunk 4]
```diff
@@ -453,6 +453,59 @@
...
+                    "--trust-remote-code",
+                    "--runtime",
+                    "ort-genai",
+                ]
+            )
+
+        assert mock_export.call_args.kwargs["trust_remote_code"] is True
+
+    def test_runtime_ort_genai_rejects_mage_vl_before_saving(self):
+        with tempfile.TemporaryDirectory() as tmpdir:
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/cli_test.py]: Fixed in 72ddb03. The temporary directory, mocks, and expected exception now share one context manager, satisfying SIM117.

**Comment by justinchuby** [comment]
### Latest-main rebase and graph optimization update

Force-pushed a linear rebase onto current `main` at `ed0319d` (Moonshine support); PR head is now `747edbf`.

The new optimization commit also:
- keeps the Mage vision graph input at processor-native `float32`, with one graph-entry cast for fp16/bf16 builds;
- derives packed-attention `cu_seqlens` from per-`grid_thw` window lengths in `O(num_visuals * max_windows)`, replacing the previous patch-by-visual metadata expansion;
- covers both packed `T=5` video input and the real processor shape (one image row plus five `T=1` video-frame rows with sampled positions through frame 719);
- adds regression assertions for fused operators and post-weight transpose folding.

#### FP16 CUDA graph evidence

Post-weight Mobius graphs:
- decoder: **45 nodes**, **2 `GroupQueryAttention`**, **4 `SkipSimplifiedLayerNormalization`**, **2 `Swish`**, **0 runtime `Transpose`**;
- vision: **197 nodes**, **2 `PackedMultiHeadAttention`**, **4 `SkipLayerNormalization`**, **0 runtime `Transpose`**, and **0 `GreaterOrEqual`** metadata-expansion nodes;
- embedding: **21 nodes**, **0 runtime `Transpose`**.

ORT `ORT_ENABLE_ALL` optimized vision graph:
- **181 nodes**;
- **2 `BiasGelu`** fusions;
- only **2 metadata `MemcpyToHost`** nodes (both scalar `Range` bounds; no patch activation CPU round-trip), down from 4 in the prior graph.

A real CUDA EP execution of the processor-shaped nonzero image+video metadata case completed with output shape `(6, 64)`, all finite, and nonzero output.

#### Validation after rebase

- `lintrunner -a`: **passed, no lint issues**
- canonical non-integration suite plus Mage parity/fusion tests: **3882 passed, 58 skipped**
- automated code-review agent: **no significant findings**

No unrelated Bamba changes were made.
